evals/solvers/providers/google/gemini_solver_wav.py

```diff
@@ -0,0 +1,239 @@
+                },
+            )
+            
+            if gen_content_resp.prompt_feedback.block_reason:
+                # Blocked by safety filters
+                solver_result = SolverResult(
+                    str(gen_content_resp.prompt_feedback),
+                    error=gen_content_resp.prompt_feedback,
+                )
+            else:
+                # Get text response
+                solver_result = SolverResult(
+                    gen_content_resp.text,
+                    error=gen_content_resp.prompt_feedback,
+                )
+        except (google.api_core.exceptions.GoogleAPIError,) as e:
+            solver_result = SolverResult(
+                e.message,
+                error=e,
+            )
+        except ValueError as e:
+            # TODO: Why does this error ever occur and how can we handle it better?
+            # (See google/generativeai/types/generation_types.py for the triggers)
+            known_errors = [
+                "The `response.text` quick accessor",
+                "The `response.parts` quick accessor",
+            ]
+            if any(err in str(e) for err in known_errors):
+                solver_result = SolverResult(
+                    str(e),
+                    error=e,
+                )
+            else:
+                raise e
+
+        return solver_result
+
+    @staticmethod
+    def _convert_msgs_to_google_format(msgs: list[Message]) -> list[GoogleMessage]:
+        """
+        Gemini API requires that the message list has
+        - Roles as 'user' or 'model'
+        - Alternating 'user' and 'model' messages
+        - Ends with a 'user' message
+        """
+        
+        # Enforce valid roles
+        gmsgs = []
+        for msg in msgs:
+            gmsg = GoogleMessage.from_evals_message(msg)
+            gmsgs.append(gmsg)
+            assert gmsg.role in {"user", "model"}, f"Invalid role: {gmsg.role}"
+        
+        # Enforce alternating messages
+        # e.g. [user1, user2, model1, user3] -> [user12, model1, user3]
+        std_msgs = []
+        for msg in gmsgs:
+            if len(std_msgs) > 0 and msg.role == std_msgs[-1].role:
+                # handle the case where msg.parts is a string or a dict. 
+                # The assumption is that std_msgs[-1].parts is a string
+                if isinstance(msg.parts[0], str):
+                    std_msgs[-1].parts = ["\n".join(std_msgs[-1].parts + msg.parts)]
+                elif isinstance(msg.parts[0], dict):
+                    std_msgs[-1].parts = [{
+                        "type": "text",
+                        "text": std_msgs[-1].parts[0] + "\n" + msg.parts[0]["text"]
+                    }, {
+                        "type": "audio_url",
+                        "audio_url": msg.parts[1]["audio_url"]
+                    }]
+
+            else:
+                # Proceed as normal
+                std_msgs.append(msg)
+
```

Comment:
I am not sure I can follow the logic in this code block. Can you provide comments on the expected msgs format? It appears that certain assumptions are made for the msg.

Reply:
I added some comments for the input format.